boringtun/src/ffi/mod.rs

```diff
@@ -160,6 +163,85 @@ pub unsafe extern "C" fn check_base64_encoded_x25519_key(key: *const c_char) ->
     }
 }
 
+/// Custom tracing_subscriber writer to an external function pointer
+struct FFIFunctionPointerWriter {
+    log_func: unsafe extern "C" fn(*const c_char),
+}
+
+/// Implements Write trait for use with tracing_subscriber
+impl Write for FFIFunctionPointerWriter {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
+        let out_str = String::from_utf8_lossy(buf).to_string();
+        if let Ok(c_string) = CString::new(out_str) {
+            unsafe { (self.log_func)(c_string.as_ptr()) }
+            Ok(buf.len())
+        } else {
+            Err(Error::new(
+                ErrorKind::Other,
+                "Failed to create CString from buffer.",
+            ))
+        }
+    }
+
+    fn flush(&mut self) -> Result<(), std::io::Error> {
+        // no-op
+        Ok(())
+    }
+}
+
+/// Sets the default tracing_subscriber to write to `log_func`.
+///
+/// Uses Compact format without level, target, thread ids, thread names, or ansi control characters.
+/// Subscribes to TRACE level events.
+///
+/// This function should only be called once as setting the default tracing_subscriber
+/// more than once will result in an error.
+///
+/// Returns false on failure.
+///
+/// # Safety
+///
+/// `c_char` will be freed by the library after calling `log_func`. If the value needs
+/// to be stored then `log_func` needs to create a copy, e.g. `strcpy`.
+#[no_mangle]
+pub unsafe extern "C" fn set_logging_function(
+    log_func: unsafe extern "C" fn(*const c_char),
+) -> bool {
+    let result = std::panic::catch_unwind(|| -> bool {
+        let writer = FFIFunctionPointerWriter { log_func };
+        let format = fmt::format()
+            // don't include levels in formatted output
+            .with_level(false)
+            // don't include targets
+            .with_target(false)
+            // don't 'include the thread ID of the current thread
+            .with_thread_ids(false)
+            // don't 'include the name of the current thread
+            .with_thread_names(false)
+            // use the `Compact` formatting style.
+            .compact()
+            // disable terminal escape codes
+            .with_ansi(false);
+
+        if let Ok(()) = fmt()
+            .event_format(format)
+            .with_writer(std::sync::Mutex::new(writer))
+            .with_max_level(tracing::Level::TRACE)
+            .with_ansi(false)
+            .try_init()
```

Comment:
If you aren't going to destructure with `if let`, use `.is_ok` instead of the `let` in the `if`.